### PR TITLE
Fixing 2D IP bug

### PR DESCRIPTION
**Pull request by jcapriot**
implements testing on apparent changeability data for the IP in 2D and 3D,
Enabled to find runtime bug for 2D IP, however accuracy of the 2D apparent changeability data seems off.

### CODE REVIEW

**Comment by jcapriot** [comment]
In relation to #976

**Comment by jcapriot** [comment]
Ok, so I've come across actually a rather large bug, when a single receiver object is used for multiple sources. It must have multiple "dc_voltage" values. This actually also replicates to a bug in the geometric factor (which is also dependent on the sources. This PR will fix both of these problems soon.

**Comment by jcapriot** [comment]
apparent_resitivity is perfectly fine to store on the receiver, but I think in the future it will be best to store dc_voltages on the simulation

**Comment by jcapriot** [comment]
This PR is now in a good state on my end to merge. @sgkang can you take a once over please?

Some highlights
* Simplification of the IP code for Jvec, JTvec, fields to rely on the parent dc simulation
* Stores geometric factor for apparent resistivity on the receivers as a dictionary indexed by the sources (allows for the same receiver to listen to multiple sources for this data_type)
* Store the _dc_voltage for apparent chargeability on the IP simulation class. This should not be on the receivers because it is not only unique to sources, but also to the mesh and conductivity model. This now allows for a single dc receiver to listen to multiple sources for apparent chargeability data.
* Updates for dcip tutorials for the new changes to discretize, instead of loading in a true_model file it's much easier to just recreate it every time. There's also some streamlining of the plotting commands in those files as well.
* Update the intersphinx mapping for matplotlib

**Comment by dccowan** [comment]
I took a look at the changes. I think this is a lot cleaner and removes some redundancies. Given that it passes all the tests (including reproducing the tutorials) I think it's good to go.

File: SimPEG/electromagnetics/static/induced_polarization/simulation.py [hunk 1]
```diff
@@ -7,35 +7,27 @@
...
-    sigma = props.PhysicalProperty("Electrical conductivity (S/m)")
-
-    rho = props.PhysicalProperty("Electrical resistivity (Ohm m)")
-
-    props.Reciprocal(sigma, rho)
+class BaseIPSimulation(BaseDCSimulation):
 
     eta, etaMap, etaDeriv = props.Invertible("Electrical Chargeability")
 
     data_type = properties.StringChoice(
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/induced_polarization/simulation.py:20]: @jcapriot: do we need this in the simulation class?

File: SimPEG/electromagnetics/static/induced_polarization/simulation.py [hunk 2]
```diff
@@ -101,145 +84,24 @@ def getJtJdiag(self, m, W=None):
...
             else:
-                W = W.diagonal() ** 2
-
-            diag = np.zeros(J.shape[1])
-            for i in range(J.shape[0]):
-                diag += (W[i]) * (J[i] * J[i])
+                W = (self._scale * W.diagonal()) ** 2
 
-            self.gtgdiag = diag
+            self.gtgdiag = np.einsum("i,ij,ij->j", W, J, J)
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/induced_polarization/simulation.py:92]: huh, that's cool. is this much more efficient?

File: SimPEG/electromagnetics/static/induced_polarization/simulation.py [hunk 3]
```diff
@@ -101,145 +84,24 @@ def getJtJdiag(self, m, W=None):
...
-
-            for rx in src.receiver_list:
-                df_dmFun = getattr(f, "_{0!s}Deriv".format(rx.projField), None)
-                df_dm_v = df_dmFun(src, du_dm_v, v, adjoint=False)
-                Jv.append(rx.evalDeriv(src, self.mesh, f, df_dm_v))
-
-        # Conductivity (d u / d log sigma) - EB form
-        # Resistivity (d u / d log rho) - HJ form
-        return self.sign * np.hstack(Jv)
+        return self._scale * super().Jvec(m, v, f)
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/induced_polarization/simulation.py:98]: nice job.

File: SimPEG/electromagnetics/static/induced_polarization/simulation_2d.py [hunk 4]
```diff
@@ -40,19 +36,18 @@ def fields(self, m):
...
+        if self._scale is None:
+            scale = Data(self.survey, np.full(self.survey.nD, self._sign))
+            f = self.fields_to_space(self._f)
             # loop through receievers to check if they need to set the _dc_voltage
             for src in self.survey.source_list:
                 for rx in src.receiver_list:
                     if (
                         rx.data_type == "apparent_chargeability"
-                        and rx._dc_voltage is None
+                        or self.data_type == "apparent_chargeability"
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/induced_polarization/simulation_2d.py:47]: again do we need this?

could be useful, if one wants to set the data type by passing the `data_type` to a simulation.

File: SimPEG/electromagnetics/static/induced_polarization/simulation_2d.py [hunk 5]
```diff
@@ -227,12 +224,12 @@ def MfRhoDerivMat(self):
         Derivative of MfRho with respect to the model
         """
         if getattr(self, "_MfRhoDerivMat", None) is None:
+            drho_dlogrho = sdiag(self.rho) * self.etaDeriv
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/induced_polarization/simulation_2d.py:227]: may be a different name `drho_dlogrho_deta_dm` ?

File: SimPEG/electromagnetics/static/resistivity/receivers.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 from ....utils.code_utils import deprecate_class
+from discretize.utils import Identity
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/resistivity/receivers.py:3]: I did not know this exists :)

File: SimPEG/electromagnetics/static/resistivity/receivers.py [hunk 7]
```diff
@@ -77,17 +73,45 @@ def eval(self, src, mesh, f):
...
+                    return v / self.geometric_factor[src][:, None]
+                return v / self.geometric_factor[src]
+            except KeyError:
+                raise KeyError(
+                    "Receiver geometric factor has not been set, please execute "
+                    "survey.set_geometric_factor()"
+                )
+        return v
+
+    def evalDeriv(self, src, mesh, f, v=None, adjoint=False):
```
Comment (sgkang) [inline SimPEG/electromagnetics/static/resistivity/receivers.py:90]: so, this is called many times. 
can we encode `geometric_factor` in `P`? (I guess this was the previous implementation).

File: tests/em/static/test_DC_2D_analytic.py [hunk 8]
```diff
@@ -330,5 +330,54 @@ def test_Simulation2DNodal(self, tolerance=0.05):
         self.assertLess(err, tolerance)
 
 
+class DCSimulationAppResTests(unittest.TestCase):
```
Comment (sgkang) [inline tests/em/static/test_DC_2D_analytic.py:333]: @jcapriot: thanks for adding the test!

File: tests/em/static/test_IP_2D_fwd.py [hunk 9]
```diff
@@ -119,5 +118,122 @@ def test_Simulation2DCellCentered(self):
         self.assertTrue(passed)
 
 
+class ApparentChargeability2DTest(unittest.TestCase):
```
Comment (sgkang) [inline tests/em/static/test_IP_2D_fwd.py:121]: nicely done!

**Review by sgkang — APPROVED** [review]
@jcapriot: Great job! and sorry for being late about this review. 
It really simplifies the IP code base, and I loved how you modified. 
Made a few minor comments, and it will be great if you can take a look. 

All the best, 

Seogi.

File: SimPEG/electromagnetics/static/induced_polarization/simulation.py (hunk 1 above)
Comment (jcapriot) [reply SimPEG/electromagnetics/static/induced_polarization/simulation.py:20]: We certainly do not, but I left it there for backwards compatibility. I should probably add a deprecation warning saying to define it on the receiver.

File: SimPEG/electromagnetics/static/induced_polarization/simulation.py (hunk 2 above)
Comment (jcapriot) [reply SimPEG/electromagnetics/static/induced_polarization/simulation.py:92]: Much more, Dom did a few tests with it to verify in #968

File: SimPEG/electromagnetics/static/induced_polarization/simulation_2d.py (hunk 4 above)
Comment (jcapriot) [reply SimPEG/electromagnetics/static/induced_polarization/simulation_2d.py:47]: Again, I think it should be more noted as deprecated behavior.

File: SimPEG/electromagnetics/static/resistivity/receivers.py (hunk 6 above)
Comment (jcapriot) [reply SimPEG/electromagnetics/static/resistivity/receivers.py:3]: There is an Identity, but I think I actually don't need to import it here...

File: SimPEG/electromagnetics/static/resistivity/receivers.py (hunk 7 above)
Comment (jcapriot) [reply SimPEG/electromagnetics/static/resistivity/receivers.py:90]: Unfortunately not, which was the reason for switching. The geometric factor is dependent on the src, so either we store a `P` for every [source, receiver] , or a `geometric_factor` for every [source, receiver] pair. This is more memory efficient with only a slight hit to computation time. This also allows for a single receiver object to be attached to many sources.

Comment (sgkang) [reply SimPEG/electromagnetics/static/resistivity/receivers.py:90]: got it! that's a good point that I did not take into account before.

**Pull request merged**